File: Discrete math/1 module/maxcomponent/MaxComponent.go
package main

import (
	"fmt"
	"github.com/skorobogatov/input"
	"strings"
)

type Component struct {
	min,edgesNum int
	vertices []int
}

func (c *Component) addVertex(vertex Vertex){
	if vertex.n<c.min{
		c.min=vertex.n
	}
	c.edgesNum+=len(vertex.next)
	c.vertices = append(c.vertices, vertex.n)
}

func (c *Component) Less(component Component)bool{
	if len(c.vertices)!= len(component.vertices){
		return len(c.vertices)< len(component.vertices)
	}
	if c.edgesNum!=component.edgesNum{
		return c.edgesNum<component.edgesNum
	}
	return c.min>component.min
}

func (c Component) hasVertex(v int) bool{
	for i:=range c.vertices{
		if c.vertices[i]==v{
			return true
		}
	}
	return false
}

func (c *Component) copy(component Component){
	c.edgesNum=component.edgesNum
	c.min=component.min
	c.vertices=c.vertices[:0]
	for i:=range component.vertices{
		c.vertices = append(c.vertices, component.vertices[i])
	}

}

type Vertex struct {
	next []*Edge
	flag bool
	n int
}

type Edge struct{
	start,end int
	flag bool
	twin *Edge
}

func dfs(vertices *[]Vertex)  *Component{
	var comp,c Component
	for i:=range *vertices{
		if !(*vertices)[i].flag{
			visitVertex(i,vertices,&c)
			if comp.Less(c) {
				comp.copy(c)
			}
			c.min=len(*vertices)
			c.edgesNum=0
			c.vertices=c.vertices[:0]
		}
	}
	return &comp
}

func visitVertex(v int,vertices *[]Vertex,c *Component){
	c.addVertex((*vertices)[v])
	(*vertices)[v].flag=true
	for i:=range (*vertices)[v].next{
		next:=(*vertices)[v].next[i]
		if !(*vertices)[next.end].flag{
			visitVertex(next.end,vertices,c)
		}
	}
}

func findComp(vertices *[]Vertex) string{
	res:=strings.Builder{}
	res.WriteString("graph {\n")
	comp:=dfs(vertices)

	for i:=range *vertices{
		res.WriteString(fmt.Sprintf("%v ",i))
		if comp.hasVertex(i){
			res.WriteString("[ color = red ]")
		}
		res.WriteString("\n")
		(*vertices)[i].flag=false
	}

	for i:=range *vertices{
		if !(*vertices)[i].flag{
			if comp.hasVertex(i){
				printEdges(&res,vertices,i,true)
			}else{
				printEdges(&res,vertices,i,false)
			}
		}
	}
	res.WriteString("}\n")
	return res.String()
}

func printEdges(res *strings.Builder,vertices *[]Vertex,v int,flag bool){
	for i:=range (*vertices)[v].next{
		next:=(*vertices)[v].next[i]
		if !next.flag{
			(*vertices)[v].next[i].flag=true
			(*vertices)[v].next[i].twin.flag=true
			res.WriteString(fmt.Sprintf("%v -- %v ",v,next.end))
			if flag{
				res.WriteString(" [color = red ]")
			}
			res.WriteString("\n")
			printEdges(res,vertices,next.end,flag)
		}
	}
}

func main(){
	var n,m int
	input.Scanf("%d %d",&n,&m)
	vertices:=make([]Vertex,n)

	for i:=0;i<m;i++{
		var a,b int
		input.Scanf("%d %d",&a,&b)
		e1:=&Edge{
			start: a,
			end:   b,
			flag:  false,
			twin:  nil,
		}
		e2:=&Edge{
			start: b,
			end:   a,
			flag:  false,
			twin:  nil,
		}
		e1.twin,e2.twin=e2,e1
		vertices[a].next = append(vertices[a].next, e1)
		vertices[b].next = append(vertices[b].next, e2)
		vertices[a].n,vertices[b].n=a,b
	}
	fmt.Println(findComp(&vertices))
}

File: Discrete math/1 module/maxcomponent/MaxComponent.java
import java.util.*;
public class MaxComponent {
    public static void main(String[] args) {
        Scanner scan = new Scanner (System.in);
        int n=scan.nextInt();
        int edges=scan.nextInt();
        if (edges==0){
            StringBuilder fin=new StringBuilder("graph{\n0 [ color = red ]\n");
            for (int i=1;i<n;i++)
                fin.append(i+"\n");
            System.out.println(fin.append("}"));
            return;
        }
        boolean[] v=new boolean[n];
        Vertex[] vertex=new Vertex[n];
        for (int i=0;i<edges;i++){
            int a=scan.nextInt();
            int b=scan.nextInt();
            if (!v[a])
                vertex[a]=new Vertex();
            v[a]=true;
            Edge e1=new Edge(a,b);
            Edge e2=new Edge(b,a);
            e1.setTwin(e2);
            e2.setTwin(e1);
            vertex[a].getList().add(e1);
            if (!v[b])
                vertex[b]=new Vertex();
            v[b]=true;
            vertex[b].getList().add(e2);
        }
        Thread t=new Thread(null,null,"Main",2<<23){
            public void run(){
                StringBuilder fin=new StringBuilder("graph{\n");
                ArrayList<Vertex> prq=DFS(vertex,v,fin);
                Collections.sort(prq);
                for (int i=0;i<prq.size();i++)
                    if (i!=prq.size()-1)
                        help(vertex,prq.get(i).getList().get(0).getFrom(),"",fin);
                    else
                        help(vertex,prq.get(i).getList().get(0).getFrom()," [ color = red ]",fin);
                System.out.println(fin.append("}\n"));
            }
        };
        t.start();


    }
    public static void help(Vertex[] vertex,int v,String s,StringBuilder fin){
        if (!vertex[v].isPrinted()){
            vertex[v].setPrinted(true);
            fin.append(v+s+"\n");
        }
        for (int i=0;i<vertex[v].getList().size();i++){
            Edge next=vertex[v].getList().get(i);
            if (!next.isPrinted()){
                next.setPrinted(true);
                next.getTwin().setPrinted(true);
                fin.append(v+"--"+next.getTo()+s+"\n");
                help(vertex,next.getTo(),s,fin);
            }
        }

    }
    public static ArrayList<Vertex> DFS(Vertex[] vertex,boolean[] v,StringBuilder fin){
        ArrayList<Vertex> prq=new ArrayList<>();
        for (int i=0,n=0;i<vertex.length;i++) {
            if (!v[i]){
                fin.append(i+"\n");
                continue;
            }
            if (!vertex[i].isMark()) {
                visitVertex(vertex, i, new Comp(n++));
                prq.add(vertex[i]);
            }
        }
        return prq;
    }
    public static void visitVertex(Vertex[] vertex,int v,Comp comp){
        if (!vertex[v].isMark()) {
            vertex[v].setMark(true);
            vertex[v].setComp(comp);
            vertex[v].getComp().inc();
        }
        if (v<vertex[v].getComp().getMin())
            vertex[v].getComp().setMin(v);
        for (int i=0;i<vertex[v].getList().size();i++){
            Edge next=vertex[v].getList().get(i);
            if (!next.isMark()) {
                next.setMark(true);
                next.getTwin().setMark(true);
                vertex[v].getComp().incEdges();
                visitVertex(vertex, next.getTo(), comp);
            }
        }
    }
}
class Vertex implements  Comparable<Vertex>{
    private ArrayList<Edge> list;
    private Comp comp;
    private boolean mark;
    private boolean printed;
    public  Vertex(){
        list=new ArrayList<>();
        mark=false;
        printed=false;
    }

    public boolean isPrinted() {
        return printed;
    }

    public void setPrinted(boolean printed) {
        this.printed = printed;
    }

    public ArrayList<Edge> getList() {
        return list;
    }

    public void setComp(Comp comp) {
        this.comp = comp;
    }

    public Comp getComp() {
        return comp;
    }

    public boolean isMark() {
        return mark;
    }

    public void setMark(boolean mark) {
        this.mark = mark;
    }

    @Override
    public int compareTo(Vertex o) {
        if (this.getComp().getCount()!=o.getComp().getCount()) {
            return this.getComp().getCount() - o.getComp().getCount();
        }
        if (this.getComp().getEdges()!=o.getComp().getEdges()) {
            return this.getComp().getEdges() - o.getComp().getEdges();
        }
        return -this.getComp().getMin()+o.getComp().getMin();
    }
}
class Comp{
    private int number;
    private int count;
    private int min;
    private  int edges;
    public Comp(int n){
        min=Integer.MAX_VALUE;
        number=n;
        edges=0;
    }
    public int getNumber() {
        return number;
    }


    public int getCount() {
        return count;
    }

    public void inc(){
        count++;
    }

    public void setMin(int min) {
        this.min = min;
    }

    public int getMin() {
        return min;
    }
    public void incEdges(){
        edges++;
    }

    public int getEdges() {
        return edges;
    }
}
class Edge{
    private boolean mark;
    private boolean printed;
    private int from;
    private int to;
    private Edge twin;
    public Edge(int a,int b){
        from=a;
        to=b;
        mark=false;
        printed=false;
    }

    public void setMark(boolean mark) {
        this.mark = mark;
    }

    public boolean isMark() {
        return mark;
    }

    public int getTo() {
        return to;
    }

    public int getFrom() {
        return from;
    }

    public void setPrinted(boolean printed) {
        this.printed = printed;
    }

    public boolean isPrinted() {
        return printed;
    }

    public Edge getTwin() {
        return twin;
    }

    public void setTwin(Edge twin) {
        this.twin = twin;
    }
}
